winch/codegen/src/abi/mod.rs

```diff
@@ -118,178 +125,310 @@ pub(crate) trait ABI {
+            ABIOperand::Reg { ty, .. } | ABIOperand::Stack { ty, .. } => ty,
         }
     }
 }
 
-/// ABI-specific representation of the function result.
-#[derive(Copy, Clone, Debug)]
-pub(crate) enum ABIResult {
-    Void,
-    Reg {
-        /// Type of the result.
-        ty: WasmType,
-        /// Register to hold the result.
-        reg: Reg,
-    },
+/// Information about the [`ABIOperand`] information used in [`ABISig`].
+#[derive(Clone, Debug)]
+pub(crate) struct ABIOperands {
+    /// All the operands.
+    pub inner: SmallVec<[ABIOperand; 6]>,
+    /// All the registers used as operands.
+    pub regs: HashSet<Reg>,
+    /// Stack bytes used by the operands.
+    pub bytes: u32,
 }
 
-impl Default for ABIResult {
+impl Default for ABIOperands {
     fn default() -> Self {
-        Self::Void
+        Self {
+            inner: Default::default(),
+            regs: HashSet::with_capacity(0),
+            bytes: 0,
+        }
     }
 }
 
-impl ABIResult {
-    /// Create a register ABI result.
-    pub fn reg(ty: WasmType, reg: Reg) -> Self {
-        Self::Reg { ty, reg }
+/// ABI-specific representation of an [`ABISig`].
+#[derive(Clone, Debug, Default)]
+pub(crate) struct ABIResults {
+    /// The result operands.
+    operands: ABIOperands,
+}
+
+impl ABIResults {
+    /// Creates [`ABIResults`] from a slice of `WasmType`.
+    pub fn from<F>(returns: &[WasmType], mut map: F) -> Self
+    where
+        F: FnMut(&WasmType, u32) -> (ABIOperand, u32),
+    {
+        if returns.len() == 0 {
+            return Self::default();
+        }
+
+        let register_capacity = returns.len().min(2);
+        let (operands, regs, bytes): (SmallVec<[ABIOperand; 6]>, HashSet<Reg>, u32) =
+            returns.iter().fold(
+                (
+                    SmallVec::new(),
+                    HashSet::with_capacity(register_capacity),
+                    0,
+                ),
+                |(mut operands, mut regs, stack_bytes), arg| {
+                    let (operand, bytes) = map(arg, stack_bytes);
+                    if operand.is_reg() {
+                        regs.insert(operand.unwrap_reg());
+                    }
+                    operands.push(operand);
+                    (operands, regs, bytes)
+                },
+            );
+
+        Self::new(ABIOperands {
+            inner: operands,
+            regs,
+            bytes,
+        })
     }
 
-    /// Createa void ABI result.
-    pub fn void() -> Self {
-        Self::Void
+    /// Create a new [`ABIResults`] from [`ABIOperands`].
+    pub fn new(operands: ABIOperands) -> Self {
+        Self { operands }
     }
 
-    /// Get the result reg.
-    pub fn result_reg(&self) -> Option<Reg> {
-        match self {
-            Self::Reg { reg, .. } => Some(*reg),
-            _ => None,
-        }
+    /// Returns an iterator over the result registers.
+    pub fn regs(&self) -> &HashSet<Reg> {
```

Comment:
Documentation is out of sync with the code: it returns a set not an iterator.

Reply:
Fixed